torch/csrc/dynamo/eval_frame.c

```diff
@@ -322,12 +342,99 @@ static void destroy_cache_entry(CacheEntry* e) {
   free(e);
 }
 
-inline static CacheEntry* get_cache_entry(PyCodeObject* code) {
-  CacheEntry* extra = NULL;
-  _PyCode_GetExtra((PyObject*)code, cache_entry_extra_index, (void*)&extra);
+/* CacheEntry helper functions ends */
+
+/* Extractions helper functions begins. They help with NULL and SKIP_CODE corner cases */
+
+inline static CacheEntry* extract_cache_entry(ExtraState* extra_state) {
+  // Helper to extra the cache_entry from the extra state.
+  if (extra_state == NULL || extra_state == SKIP_CODE) {
+    return NULL;
+  }
+  return extra_state->cache_entry;
+}
+
+
+inline static FrameState* extract_frame_state(ExtraState* extra_state) {
+  // Returns either the previously stored frame state or an empty dict.
+  if (extra_state == NULL || extra_state == SKIP_CODE) {
+    return NULL;
+  }
+  return extra_state->frame_state;
+}
+
+/* Extractions helper functions ends */
+
+
+/* Extra state helper functions begins */
+
+inline static ExtraState* get_extra_state(PyCodeObject* code) {
+  ExtraState* extra = NULL;
+  _PyCode_GetExtra((PyObject*)code, extra_index, (void*)&extra);
   return extra;
 }
 
+inline static void set_extra_state(PyCodeObject* code, ExtraState* extra_state) {
+  // Sets the extra state on the extra scrach space of the code object.
+  _PyCode_SetExtra((PyObject*)code, extra_index, extra_state);
+}
+
+static ExtraState* create_extra_state(
+    CacheEntry* cache_entry,
+    FrameState* frame_state) {
+  // Creates a new extra state with the given cache_entry and frame_state.
+  // It steals the reference for both cache_entry and frame_state. These are
+  // freed in destroy_extra_state.
+  ExtraState* extra_state = (ExtraState*)malloc(sizeof(ExtraState));
+  DEBUG_NULL_CHECK(extra_state);
+  extra_state->cache_entry = cache_entry;
+  extra_state->frame_state = frame_state;
+  return extra_state;
+}
+
+
+inline static void create_and_set_extra_state(
```

Comment:
invariant: there must not be any pre-existing extra state on code

Reply:
Added